graphql/src/store/prefetch.rs

```diff
@@ -0,0 +1,752 @@
+        &self.entity
+    }
+}
+
+impl Node {
+    fn typename(&self) -> &str {
+        self.get("__typename")
+            .expect("all entities have a __typename")
+            .as_str()
+            .expect("__typename must be a string")
+    }
+}
+
+/// Encapsulate how we should join a list of parent entities with a list of
+/// child entities.
+#[derive(Debug)]
+struct Join<'a> {
+    /// The object type of the child entities
+    child_type: ObjectOrInterface<'a>,
+    /// The name of the field in the parent type
+    parent_field: &'a str,
+    /// Whether the parent field is a list or scalar
+    parent_is_list: bool,
+    /// The name of the field in the child type
+    child_field: &'a str,
+    /// Whether the child field is a list or a scalar
+    child_is_list: bool,
+}
+
+impl<'a> Join<'a> {
+    /// Construct a `Join` based on the parent field pointing to the child
+    fn new(schema: &'a s::Document, field: &'a s::Field) -> Self {
+        let child_type = object_or_interface_from_type(&schema, &field.field_type)
+            .expect("we only collect fields that are objects or interfaces");
+
+        let (parent_field, parent_is_list, child_field, child_is_list) =
+            if let Some(derived_from_field) = sast::get_derived_from_field(child_type, field) {
+                let is_list = sast::is_list_or_non_null_list_field(derived_from_field);
+                ("id", false, derived_from_field.name.as_str(), is_list)
+            } else {
+                let is_list = sast::is_list_or_non_null_list_field(field);
+                (field.name.as_str(), is_list, "id", false)
+            };
+        Join {
+            child_type,
+            parent_field,
+            parent_is_list,
+            child_field,
+            child_is_list,
+        }
+    }
+
+    /// Perform the join. The child nodes are distributed into the parent nodes
+    /// according to the join condition, and are stored in the `response_key`
+    /// entry in each parent's `children` map.
+    ///
+    /// The `children` must contain the nodes in the correct order for each
+    /// parent; we simply pick out matching children for each parent but
+    /// otherwise maintain the order in `children`
+    fn perform(&self, parents: &mut Vec<Node>, children: Vec<Node>, response_key: &str) {
+        let children: Vec<_> = children.into_iter().map(|child| Rc::new(child)).collect();
+
+        if is_root_node(parents) {
+            let root = parents
+                .first_mut()
+                .expect("is_root_node checked that we have a node");
+            root.children.insert(response_key.to_owned(), children);
+            return;
+        }
+
+        // Build a map child_key -> Vec<child> for joining
+        let mut grouped: BTreeMap<&str, Vec<Rc<Node>>> = BTreeMap::default();
+        for child in children.iter() {
+            match child
+                .get(self.child_field)
+                .expect("child must have a value in child_field")
```

Comment:
The `expect` message is misleading: what we expect is that `child` has an entry under the key `child_field`; that must be true because we got children by running a query like `child_field in (..)` So there shouldn't be any chance to crash, since we wouldn't get a child without an entry for `child_field` from the database (for relational storage, the query would fail with an error; for JSONB storage, the query will skip rows where `data->$child_field` does not exist)

I improved the `expect` message